coinbaseadvanced/client.py

```diff
@@ -557,6 +557,44 @@ def get_product_candles(
         product_candles = CandlesPage.from_response(response)
         return product_candles
 
+    def get_product_candles_all(
+           self,
+            product_id: str,
+            start_date: datetime,
+            end_date: datetime,
+            granularity: Granularity) -> CandlesPage:
+        """
+        Gets all requested product candles
+        """
+
+        # pre-calculate 300 granularity entries in minutes 
+        minutesX300 = Gran[granularity.value] * 300
+
+        # run through from most recent to oldest
+
+        loop_end_date = end_date
+        # calculate start date for 300 entries (max allowed by coinbase)
+        loop_start_date = end_date - timedelta(minutes=minutesX300)
+        # avoid asking for more than requested
+        if loop_start_date < start_date:
+            loop_start_date = start_date
+        # get the initial batch of candles
+        product_candles = self.get_product_candles(product_id, loop_start_date, loop_end_date, granularity)
+
+        # while we still have not gotten all the requested candles loop until all are requested
+        while loop_start_date > start_date:
+            # offset end by one granularity to avoid duplicates
+            loop_end_date = loop_start_date - timedelta(minutes=Gran[granularity.value])
+            # recalculate start for the previous (older) 300 candles
+            loop_start_date = loop_end_date - timedelta(minutes=minutesX300)
+            # avoid asking for more than requested
+            if loop_start_date < start_date:
+                loop_start_date = start_date
+            # get the next batch and extend the list
+            product_candles.candles.extend(self.get_product_candles(product_id, loop_start_date, loop_end_date,granularity).candles)
+        
+        return product_candles
```

Comment:
Overall, I think this looks good, and I like your idea about reusing the `CandlesPage` here. I can see some logic duplication with what you have outside/inside of the loop; what do you think if we refactor that a little bit?

See below some code about how potentially we can make this logic more compact (take it as a suggestion):

```
def get_product_candles_all(product_id:str, start_date:datetime, end_date:datetime, granularity:Granularity) -> List[Candle]:
  
  granularity_in_mins = GRANULARITY_MAP_IN_MINUTES[granularity.value]
  limit = 300 # [Coinbase API enforcement] Max amount of candles that can be returned.

  candles = list()

  dt_cursor = start_date
  while(dt_cursor < end_date):
    next_dt = dt_cursor + timedelta(minutes=granularity_in_mins*(limit-1))
    if(next_dt >= end_date):
      next_dt = end_date

    candles_page = client.get_product_candles(product_id="ALGO-USD", start_date=dt_cursor, end_date=next_dt, granularity=granularity)
    candles.extend(candles_page.candles)
    
    dt_cursor = next_dt + timedelta(minutes=granularity_in_mins)

  candles.sort(reverse=True, key=lambda c: c.start)
  return candles
```

We can introduce your idea here about reusing the `CandlesPage` object. Also, we'll probably benefit from adding a unit test for this endpoint. What do you think?

Reply:
As I mentioned before, I'm new to python, so suggest away.
Most of my coding experience is ages old, and focused on embedded ASM programming. 

I'm guessing that your GRANULARITY_MAP_IN_MINUTES replaces my GRAN in products.py?

In your client.get_product_candles call you override the product_id? A typo perhaps?

Your refactor changes the direction of the calls to get_product_candles. The reason I went from end_date, paging back to start_date is because that is the order coinbase returns them. Most recent candle first. By calling in reverse order and extending, the list stays sorted. I see you run a sort of the candles before returning, but I'd prefer eliminating the performance hit. That is also why I left the multiply of 300 * granularity out of the loop.

I haven't tried calling get_product_candles with an end_date in the future, so I'm not certain how coinbase will handle your next_dt being potentially a future date. 

Your code does look more concise, I like it.

As for the unit test, I'll have to do some research.

Reply:
> I'm guessing that your GRANULARITY_MAP_IN_MINUTES replaces my GRAN in products.py?
yup.

> In your client.get_product_candles call you override the product_id? A typo perhaps?
Yes, it's just a typo. You can take the example I posted more like a suggestion/guidance/template instead of exactly as-is.

> By calling in reverse order and extending, the list stays sorted. I see you run a sort of the candles before returning, but I'd prefer eliminating the performance hit.
Yes, I can reason with that, feel free to update the code I shared with your way.

Reply:
> I'm guessing that your GRANULARITY_MAP_IN_MINUTES replaces my GRAN in products.py?

yup.

Reply:
> In your client.get_product_candles call you override the product_id? A typo perhaps?

Yes, it's just a typo. You can take the example I posted more like a suggestion/guidance/template instead of exactly as-is.

Reply:
> By calling in reverse order and extending, the list stays sorted. I see you run a sort of the candles before returning, but I'd prefer eliminating the performance hit.

Yes, I can reason with that, feel free to update the code I shared.